File: 8.string-to-integer-atoi.cpp
/*
 * @lc app=leetcode id=8 lang=cpp
 *
 * [8] String to Integer (atoi)
 */

// @lc code=start
#include <string>
#include <cmath>
#include <climits>

using std::string;

class Solution
{
public:
    int myAtoi(string s)
    {
        int i = 0;
        int n = s.size();
        int int_min = INT_MIN;
        int int_max = INT_MAX;

        while (i < n && s[i] == ' ')
            i++;

        if (i == n)
            return 0;

        int sign = 1;
        if (s[i] == '+')
        {
            i++;
        }
        else if (s[i] == '-')
        {
            sign = -1;
            i++;
        }

        int num = 0;
        while (i < n && isdigit(s[i]))
        {
            int digit = s[i] - '0';
            if (num > (int_max - digit) / 10)
            {
                return sign == 1 ? int_max : int_min;
            }

            num = num * 10 + digit;
            i++;
        }

        return sign * num;
    }
};
// @lc code=end

File: 8.string-to-integer-atoi.py
#
# @lc app=leetcode id=8 lang=python3
#
# [8] String to Integer (atoi)
#

# @lc code=start
class Solution:
    def myAtoi(self, s: str) -> int:
        i = 0
        n = len(s)
        INT_MIN, INT_MAX = -2**31, 2**31 - 1

        while i < n and s[i] == " ":
            i += 1

        if i == n:
            return 0

        sign = 1
        if s[i] == "+":
            i += 1
        elif s[i] == "-":
            sign = -1
            i += 1

        num = 0
        while i < n and s[i].isdigit():
            digit = int(s[i])

            if num > (INT_MAX - digit) // 10:
                return INT_MAX if sign == 1 else INT_MIN

            num = num * 10 + digit
            i += 1

        return sign * num

# @lc code=end

